API,ENH: Allow forcing an array result in ufuncs

seberg opened the pull request:
As of now a WIP, because it would be nice to see what `_methods.py`.  *Happy if someone just picks it up also, thought I would show how trivial it is now.*

The API change is to allow `out=...`, for multi-argument outputs it is currently `out=(..., ...)`, although I guess we should allow short-handing that (nobody needs to pick it per argument).

I.e. all of these return arrays (and unlike the last time around it works with subclasses, although right now it passes `...` as context so that may confuse them or `__array_ufunc__` implementations.):
```
np.add(1, 2, out=...)
a = np.add.reduce(np.arange(10), out=...)
```

There is some discussion in gh-14489 about this.  I still think `...` is the best, because `out=np.ndarray` doesn't generalize for `__array_ufunc__`, and yes, I like `...` because it also indicates an array return in indexing.  (`arr[i, ...]` always returns an array.)

---

The ufunc part works (not sure if 100% complete), but the `_methods.py` changes do not (and change behavior subtlely, if possibly correctly). The actual change is trivial (after I had refactored this a while ago).

I think we should do this, but the fact that the `_methods.py` changes are not trivial makes me think there may be some idea missing.

It may be that the solution is for these functions to use conversion helper once at the start and then convert to scalars based on that at the end though (rather than dragging around subclasses).

---

CC @mdhaber you may be interested in this.  We could probably just put this through (with a few new tests) without the `_methods.py` changes hashed out.  (There must be a few other low-hanging places where this would simplify code, that are not as maze-like as `_methods.py`.)

seberg committed:
API,ENH: Allow forcing an array result in ufuncs

As of now a WIP, because it would be nice to see what `_methods.py`.

The API change is to allow `out=...`, for multi-argument outputs it
is currently `out=(..., ...)`, although I guess we should allow
short-handing that (nobody needs to pick it per argument).

There is some discussion in gh-14489 about this.  I still think `...`
is the best, because `out=np.ndarray` doesn't generalize for
`__array_ufunc__`, and yes, I like `...` because it also indicates an
array return in indexing.  (`arr[i, ...]` always returns an array.)

---

The ufunc part works (not sure if 100% complete), but the `_methods.py`
changes do not (and change behavior subtlely, if possibly correctly).
The actual change is trivial (after I had refactored this a while ago).

I think we should do this, but the fact that the `_methods.py` changes
are not trivial makes me think there may be some idea missing.

It may be that the solution is for these functions to use conversion
helper once at the start and then convert to scalars based on that at the
end though (rather than dragging around subclasses).

jorenham commented:
I can type it, so I like it 👌🏻

ngoldbaum commented:
Thanks for working on this! I think having a way to opt in to this behavior is the first and necessary step to eventually getting rid of scalars. Worth experimenting with for sure.

seberg commented:
> Thanks for working on this! I think having a way to opt in to this behavior is the first and necessary step to eventually getting rid of scalars.

Well, I have all the code to ensure that 0-D arrays in returns 0-D arrays (and reductions return scalars for `axis=None`).
The problem with that isn't making it work with NumPy.  The main problem by now is seeing what it means for downstream, because they would start returning arrays (if they use `np.asarray()` even where a NumPy ufunc would not).

seberg commented:
But yes, I agree... No matter how we evolve, having `out=...` for now won't stand in the way and help transitioning (as well as being immediately helpful).

mhvk commented on numpy/_core/_methods.py:
```diff
@@ -168,7 +160,8 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False, *,
         # not be either.
-        arrmean = umr_sum(arr, axis, dtype, keepdims=True, where=where)
+        arrmean = umr_sum(arr, axis, dtype, keepdims=True, where=where,
+                          out=Ellipsis)
```
Why not just `out=...`?

seberg replied:
Yeah, I think some old habit of mine when `...` wasn't actually allowed everywhere.

Anyway, maybe I should just undo the changes here.  Thought I would share them, but the details are pretty tricky for `_methods.py` as mentioned.
So it would be much easier to skip it and just focus on documentation and the few super simple places.

seberg replied:
Anyway, decided to revert this.  Need to add new tests, but I think this is useful even without the `_methods.py` fixes/changes, so let's untangle them.

mhvk commented on numpy/_core/_methods.py:
```diff
@@ -202,34 +191,25 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False, *,
 
-    ret = umr_sum(x, axis, dtype, out, keepdims=keepdims, where=where)
+    ret = umr_sum(
+        x, axis, dtype, Ellipsis if out is None else out,
```
Here too just use `...`

mhvk commented on numpy/_core/_methods.py:
```diff
@@ -132,19 +132,11 @@ def _mean(a, axis=None, dtype=None, out=None, keepdims=False, *, where=True):
-        ret = ret / rcount
+    ret = umr_sum(arr, axis, dtype, ... if out is None else out, keepdims, where=where)
+    # TODO: dtype=type(ret.dtype) castst first and isn't quite ideal that way.
+    ret = um.true_divide(ret, rcount, out=out, dtype=type(ret.dtype), casting='unsafe')
```
This should be `out=ret` by just following the array stanza of the above. Is the `dtype` then still needed?

mhvk commented on numpy/_core/_methods.py:
```diff
@@ -202,34 +191,25 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False, *,
-
+    # TODO: dtype=ret.dtype castst first and isn't quite ideal that way.
+    ret = um.true_divide(
+            ret, rcount, out=out, casting='unsafe', dtype=type(ret.dtype))
```
As above, why not `out=ret`?

seberg replied:
This shows one of the trickier things... If we want to keep allowing a scalar return here (which we must), then `out=ret` in the last step doesn't work always.  Although, you could use it in some cases.

The solution is maybe really best as `conv = array_converter(...)` and then use `conv.wrap()` at the end, although it'll change the wrapping slightly (i.e. _not_ happening inside ufuncs)...  And that might break `np.ma.` or so...

seberg replied:
Yeah, I did this to be more sure it was correct(er) for starters.  I had some test failures related to `subclass[()]` going to scalar but the ufunc path (array-wrap) ignoring the `return_scalar=True`.

mhvk commented on numpy/_core/_methods.py:
```diff
@@ -202,34 +191,25 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False, *,
+    ret = _var(a, axis=axis, dtype=dtype, out=out if out is not None else Ellipsis,
+               ddof=ddof, keepdims=keepdims, where=where, mean=mean)
 
+    ret = um.sqrt(ret, out=out)
```
And here again, `out=ret`

mhvk reviewed (commented):
Nice to see this revived! Some transcription errors, though! (see inline). Also, I'd tend to use `...` everywhere it is allowed instead of `Ellipsis`.

I guess a more general question is whether it makes sense to have a special name for this (lust like `np.new_axis is None`), maybe `np.keep_as_array`. Although, while `None` has no obvious connection to adding an axis, `...` has precedent for indexing (as you note), so on balance it seems fine as is.  Obviously, do need to add documentation as well.

seberg committed:
Revert method.py changes for now (they are just trickier)

seberg committed:
Tighten when `out=...` is allowed

This also means we don't pass it out as a context, which is probably
safer.  (it is passed as part of `return_scalar=True` already)

seberg committed:
DOC: Add release note fragment for `out=...`

Signed-off-by: Sebastian Berg <sebastianb@nvidia.com>

mhvk commented on numpy/_core/src/umath/ufunc_object.c:
```diff
@@ -3557,7 +3563,11 @@ PyUFunc_GenericReduction(PyUFuncObject *ufunc,
-        if (_set_full_args_out(1, out_obj, &full_args) < 0) {
+        if (out_obj == Py_Ellipsis) {
+            out_obj = NULL;
+            return_scalar = 0;
```
Since you use `NPY_FALSE` below, maybe here too?

mhvk commented on numpy/_core/src/umath/ufunc_object.c:
```diff
@@ -3487,6 +3487,7 @@ PyUFunc_GenericReduction(PyUFuncObject *ufunc,
      */
     PyObject *otype_obj = NULL, *out_obj = NULL, *indices_obj = NULL;
     PyObject *keepdims_obj = NULL, *wheremask_obj = NULL;
+    int return_scalar = 1;  /* scalar return is disabled for out=... */
```
Here also use `NPY_TRUE`, like below?

mhvk reviewed (commented):
Yes, fine to start with the simple case. For documentation, I guess the main place is `_add_newdocs.py` for the various `ufunc` cases (starting around l.4900). And then the tests of course...

seberg committed:
Add basic tests for ufunc out=...

seberg committed:
DOC: Add docs about `out=...` to the ufunc docs

seberg committed:
expand test a bit more to cover subclasses

(in a sense, I care less about subclasses/arrays because I think
we jus tshouldn't return scalars there, but..._

seberg commented:
Going to mark this ready for review.  It feels a bit light on documentation, so we can merge, but we may want to search if there is some other place to document this (or maybe broadcast it out through some other channels...).

mdhaber commented:
Thanks @seberg. Happy to see movement in this direction!

mhvk committed:
DOC: clarify ufunc docstrings for new out=... possibility

mhvk committed:
DOC: update user/reference docs for new out=... possibility

mhvk reviewed (approved):
@seberg - this looks all good, but I thought maybe the doc changes could be a bit better. I mention these inline, but after doing so, I felt that this was just as likely to lead to more delay. So, instead, I force-pushed the simpler change to your branch, and will open another one with a bit of docs follow-up.  Hope that makes sense!  With that, this one is all OK as far as I'm concerned.

seberg commented:
> So, instead, I force-pushed the simpler change to your branch

Thanks! Nothing nicer than seeing pushes to your branch (in my book) :).  I'll wait for that and bring it up in the meeting tonight, but maybe we can put it in very soon then.

mhvk commented:
@seberg - my follow-up PR is at https://github.com/seberg/numpy/pull/51 - this is to your branch just so can look at it. If you think it is uncontroversial, do feel free to include it in your PR. But happy to do it as follow-up too.

ngoldbaum commented:
I think this feature probably deserves at least a paragraph somewhere on this page: https://numpy.org/doc/stable/user/basics.ufuncs.html

seberg commented:
> I think this feature probably deserves at least a paragraph somewhere on this page:

Yes, Marten added that in his PR to mine, just need to manage to merge it (the github API doesn't let me merge it into my branch though, right now...)

seberg committed:
Merge pull request #51 from mhvk/ufunc-array-out-doc-followup

Ufunc array out doc followup

ngoldbaum commented on doc/release/upcoming_changes/28576.new_feature.rst:
```diff
@@ -0,0 +1,15 @@
+This is especially problematic for non-numerical dtypes (e.g. ``object``).
+
+For ufuncs (i.e. most simple math functions) it is now possible
+to use ``out=...`` which is identical in behavior to no output
```
Just to make this painfully clear, also worded a little more clearly IMO

```suggestion
to use ``out=...`` (literally `...`, e.g. ``out=Ellipsis``) which is identical in behavior to `out` not
```

ngoldbaum commented on doc/source/reference/ufuncs.rst:
```diff
@@ -49,16 +49,15 @@ output (which can be None for arrays to be allocated by the ufunc).
+If 'out' is None (the default), a uninitialized output array is created,
+which will be filled in the ufunc.  At the end, this array is returned
+unless it is zero-dimensional, in which case it is converted to a scalar;
+this conversion can be avoided by passing in ``out=...``.
```
```suggestion
this conversion can be avoided by passing in ``out=...``. This can also be 
spelled `out=Ellipsis` if you think that is clearer.
```

ngoldbaum commented on doc/source/user/basics.ufuncs.rst:
```diff
@@ -103,10 +103,16 @@ of the previous operation for that item.
+:class:`ndarrays <numpy.ndarray>`, then the output will be as well.
+The exception is when the result is zero-dimensional, in which case the
+output will be converted to an `array scalar <arrays.scalars>`. This can
+be avoided by passing in ``out=...``.
```
```suggestion
be avoided by passing in ``out=...`` or ``out=Ellipsis``.
```

ngoldbaum reviewed (commented):
Just some doc suggestions related to `...` being visually (albeit not syntactically) ambiguous, particularly in a terminal without rst rendering.

Overall looks great, this is a neat new feature that I'm sure array API implementers might be interested in.

Did this get a mailing list ping? It probably deserves one.

seberg committed:
Apply suggestions from code review

Co-authored-by: Nathan Goldbaum <nathan.goldbaum@gmail.com>

seberg commented:
> I'm sure array API implementers might be interested in.

Maybe interesting, not obviously/immediately useful. For that you would have to make that namespace actually use it, which is possible but not as long as it is the identical namespace.
(There is also always the worry that e.g. SciPy might actually want to return scalars to look like NumPy in some places...  Although, my opinion there is that SciPy should maybe lead and avoid scalars, not NumPy)

lucascolley commented:
> I'm sure array API implementers might be interested in.

👍!

> Maybe interesting, but not obviously/immediately useful, I think. For that you would have to make that namespace actually use it, which is possible but not as long as it is the identical namespace.

Relatively immediate I think. In SciPy and scikit-learn we aren't using the `numpy` namespace proper, everything goes through `array_api_compat.numpy` for code that has been converted to support the standard. So I think it would just need a PR to array-api-compat to map `array_api_compat.numpy.sum` to `numpy.sum(..., out=...)`.

> There is also always the worry that e.g. SciPy might actually want to return scalars to look like NumPy in some places...

Yes, but we already have the pattern `return x[()] if x.ndim == 0 else x` about 50 times for places where we want to return scalars for this reason but may not end up using a ufunc (or things otherwise end up coerced back to 0D arrays). So I don't think this PR will make things tangibly worse, apart from maybe needing to use that pattern in some extra places. Our test suite should be ready for this, we have assertions which forbid 0D NumPy arrays at https://github.com/scipy/scipy/blob/main/scipy/_lib/_array_api_no_0d.py.

> my opinion there is that SciPy should maybe lead and avoid scalars, not NumPy

We already do in some parts of SciPy, like `special.factorial` and `ndimage` (see https://github.com/scipy/scipy/blob/0f1fd4a7268b813fa2b844ca6038e4dfdf90084a/scipy/ndimage/_support_alternative_backends.py#L39-L41). But the general vibe I get from https://github.com/scipy/scipy/issues/18768 is that it seems very difficult to make a full transition in SciPy without NumPy shifting on things like

```python
np.array(0) * 2     # scalar, not 0d array
- np.array(0)       # scalar, not 0d-array
```

seberg commented:
Right, if you do your compat namespace it is useful.  I am fine with you doing, but it does basically cement that you cannot plan to *ever* get away from that.
To be fair, I don't mind that, I just feel if that is so, then maybe NumPy should export the namespace like that and drop the pretense that it is identical.

lucascolley commented:
> Relatively immediate I think

(Modulo difficulties with supporting this for older np versions)

miccoli commented on doc/release/upcoming_changes/28576.new_feature.rst:
```diff
@@ -0,0 +1,15 @@
+For ufuncs (i.e. most simple math functions) it is now possible
+to use ``out=...`` (literally `...`, e.g. ``out=Ellipsis``) which is identical in behavior to ``out`` not
+being passed, but will ensure a non-scalar return.
+This spelling is borrowed from ``arr1d[0, ...]`` where the ``...``
+also ensures a non-scalar return.
```
Maybe we could also mention that `arr0d[...]` is a valid lvalue as in `arr0d[...] = 3.14`

seberg replied:
TBH, I don't really understand why `arr[...]` assignment is relevant in this context?  (Besides any excuse for telling folks about `...` is a good excuse ;).)

miccoli replied:
You are right, my comment is not relevant.

From the observation that `arr0d[()]` is scalar and `arr0d[...]` is a 0D array I *wrongly* assumed that the first is not a valid lvalue... 

But I just realized that `arr0d.__setitem__((), 3.14)` and `arr0d.__setitem__(Ellipsis, 3.14)` are prefectly equivalent, so please forget my comment.

seberg commented:
Since there were no comments on the mailing list, and I don't think it's a huge change (with an approval), I'll plan to merge this in a few days unless there are comments.

seberg commented:
This summarizes most of it: https://github.com/numpy/numpy/pull/28576#issuecomment-2748619482, I have repeated my opinions on this a few times recently...

> I understand that won't happen any time soon, but NumPy 3.0 perhaps??

Maybe, but I am personally **not** willing to look into getting rid of scalars entirely (just explaining my opinion here over and over is enough ;)).
I am simply not fully convinced it is the right/feasible:
1. The pain we feel _doesn't_ come from scalars (yes it causes implementation pain).  It origiates from converting 0-D arrays to scalars so they sneaking up like ninjas.
2. Scalars do have some advantages:
   * They are not containers (and shouldn't pretend to be).  If they are, they may be very different containers (e.g. strings).
   * Scalars are e.g. immutable.
   * speed
3. I am not convinced that the transition wouldn't be utterly horrible and devastating until someone tries around a bit, and I won't be the one to do it.  For example, I could imagine that `pandas` would require a lot of work if you were to do it.

So, the thing I have already is a branch that stop converting to 0-D arrays while returning scalars where it makes sense to do so.  (See other issue.)
I can also see DTypes deciding they don't have a scalar (effectively), or even a context manager to enable _this_ or enable my other branch optionally if things are hard...
The problem with that branch isn't actually so much NumPy, it is more figuring out what SciPy needs to do to align itself (in whatever way it wants), or what we need to provide to allow that.

Apparently, I haven't yet pointed to there from this particular PR, but this discussion should be in: https://github.com/numpy/numpy/issues/24897 or similar issues.

seberg commented:
OK, thanks everyone, I'll put this in also because I suspect it is relevant to gh-28624.
